partiql-eval/src/main/kotlin/org/partiql/eval/internal/helpers/ToBag.kt

```diff
@@ -1,18 +1,35 @@
 package org.partiql.eval.internal.helpers
 
-import org.partiql.value.BagValue
-import org.partiql.value.PartiQLValue
-import org.partiql.value.PartiQLValueExperimental
+import org.partiql.errors.TypeCheckException
+import org.partiql.eval.value.Datum
+import org.partiql.types.PType
 
 /**
- * Coercion function F for bag operators described in RFC-0007
+ * Coercion function F for bag operators described in RFC-0007.
  *  - F(absent_value) -> << >>
  *  - F(scalar_value) -> << scalar_value >> # singleton bag
  *  - F(tuple_value)  -> << tuple_value >>  # singleton bag, see future extensions
  *  - F(array_value)  -> bag_value          # discard ordering
  *  - F(bag_value)    -> bag_value          # identity
+ *
+ * Throws a [TypeCheckException] when coercing a non-collection to a bag.
  */
-@OptIn(PartiQLValueExperimental::class)
-internal fun PartiQLValue.toBag(): BagValue<*> {
-    TODO("For OUTER set operators")
+internal fun Datum.asIterator(coerce: Boolean): Iterator<Datum> {
```

Comment:
(self-review) coercion behavior is slightly different than other type check exceptions in permissive mode. For the bag operators, we coerce to a bag rather than return missing.